src/tls/wsinstance-factory.c

```diff
@@ -0,0 +1,151 @@
+    return 0;
+
+  /* This is our job. */
+  debug (FACTORY, "  -> sending result.");
+  send_all (SD_LISTEN_FDS_START, result, strlen (result), 5 * 1000000);
+  *job_path = NULL;
+
+  return 0;
+}
+
+int
+main (void)
+{
+  char fingerprint[64 + 1];
+  sd_bus_error error = SD_BUS_ERROR_NULL;
+  sd_bus *bus = NULL;
+  char unit[UNIT_MAX + 1];
+  sd_bus_message *reply = NULL;
+  const char *job_path = NULL;
+  char **fdnames;
+  int r;
+
+  if (sd_listen_fds_with_names (false, &fdnames) != 1 || strcmp (fdnames[0], "connection") != 0)
+    errx (1, "Must be spawned from a systemd service on a socket with Accept=yes %s", fdnames[0]);
+
+  if (!recv_all (SD_LISTEN_FDS_START, fingerprint, sizeof fingerprint, 10 * 1000000))
+    errx (1, "Didn't receive fingerprint");
+
+  r = sd_bus_open_system (&bus);
+
+  if (r < 0)
+    errx (1, "Failed to connect to system bus: %s", strerror (-r));
+
+  /* We use the job_path variable to communicate with the match function
+   * in two ways:
+   *
+   *  - we set it to the path of the job that we're waiting to exit so
+   *    that the match function knows which signal is for us
+   *
+   *  - once the job is removed, the match function clears the variable
+   *    back to NULL.  that's how we know when to stop waiting.
+   *
+   * In effect, the duration of job_path being set to non-NULL is more
+   * or less equal to the duration of the existance of a job object at
+   * that path.
+   */
+  r = sd_bus_match_signal_async (bus, NULL,
+                                 "org.freedesktop.systemd1", "/org/freedesktop/systemd1",
+                                 "org.freedesktop.systemd1.Manager", "JobRemoved",
+                                 match_job_removed, NULL, &job_path);
+
+  if (r < 0)
+    errx (1, "Failed to install match rule: %s", strerror (-r));
+
+  /* can't fail, because fingerprint is small */
+  r = snprintf (unit, sizeof unit, "cockpit-wsinstance-https@%s.socket", fingerprint);
+  assert (r < sizeof unit);
+
+  debug (FACTORY, "Requesting start of unit %s", unit);
+  r = sd_bus_call_method (bus,
+                          "org.freedesktop.systemd1", "/org/freedesktop/systemd1",
+                          "org.freedesktop.systemd1.Manager", "StartUnit",
+                          &error, &reply, "ss", unit, "replace");
+
+  if (r < 0)
+    errx (1, "Method call failed: %s", error.message);
+
+  r = sd_bus_message_read (reply, "o", &job_path);
+
+  if (r < 0)
+    errx (1, "Invalid message response: %s", strerror (-r));
+
+  debug (FACTORY, "  -> job is %s", job_path);
+  debug (FACTORY, "Waiting for signal.");
+
+  while (sd_bus_process (bus, NULL))
+    ;
+
+  struct timespec start = { 0, 0 };
+  uint64_t remaining;
+  while (job_path && get_remaining_timeout (&start, &remaining, 20 * 1000000))
+    {
+      debug (FACTORY, "sd_bus_wait(%llu)", (long long) remaining);
+      r = sd_bus_wait (bus, remaining);
+      if (r < 0)
+        errx (1, "Error while waiting for bus: %s", strerror (-r));
```

Comment:
Thinking aloud: This could return with `r == 0`, i. e. it timed out without seeing a result. This should be okay though, as the following sd_bus_process would then just not do anything.